src/sentry/static/sentry/app/types/index.tsx

```diff
@@ -363,3 +366,29 @@ export type SentryApp = {
     slug: string;
   };
 };
+
+export type PermissionValue = 'no-access' | 'read' | 'write' | 'admin';
+
+export type Permissions = {
+  Event: PermissionValue;
+  Member: PermissionValue;
+  Organization: PermissionValue;
+  Project: PermissionValue;
+  Release: PermissionValue;
+  Team: PermissionValue;
+};
+
+//See src/sentry/api/serializers/models/apitoken.py for the differences based on application
+type BaseApiToken = {
```

Comment:
doesn't this also have `application: null | string` ?

Reply:
@MeredithAnya Good questions. At the serializer model level, that is true. But we should know if the application is null or a string in the right context. Hence why I made a base type `BaseApiToken` and extended it with `ApiTokenNoApp` for the case where `application=null`. This way, I don't have to put conditional checks for `token` when I know it will be there in the context of Sentry Applications tokens. We can make the type `ApiTokenWithApp` when `application` is not null, but right now we don't need it.

Reply:
> We can make the type `ApiTokenWithApp` when application is not null, but right now we don't need it.

Cool yeah. +1 for this in the future

Reply:
You can handle this difference at the type level instead of the type name.
You can do something like:

```
export type ApiToken = BaseApiToken & ({
  application: null;
  token: string;
  refreshToken: string;
} | {application: string});
```

Reply:
@billyvg Tried that but doesn't work. Unless I actually check the value of `application` to see if it is null, it's indeterminant if `token` is there and TS complains.

![Screen Shot 2019-09-05 at 10 01 23 AM](https://user-images.githubusercontent.com/8533851/64362936-4a623300-cfc4-11e9-834d-0e351b46e20f.png)

Reply:
Ah, just took a look at the serializer, since they are mutually exclusive, can we do:

```
export type ApiToken = BaseApiToken & ({
  token: string;
  refreshToken: string;
} | {application: string});
```

Reply:
@billyvg Still the same error. Better to have two different types if you know which is valid in each scenario. Your type works better if you don't know which version you'll get.

Reply:
👍 	- Is there a name we could use that doesn't describe its structure? e.g. ApplicationApiToken ... idk what the non application is called

Reply:
@billyvg I think there are two cases where we leave out the `application`:

1. Tokens for internal integrations
2. Tokens that are not associated with an integration such as a user token

So coming up with a name that works for both is kind of tough...